### ISSUE TITLE
/core/packages should not be visible to public

### ISSUE THREAD

**Issue by mdxthemes** [modxcms/revolution#12490]
/core/packages should not be visible to public. I can simply put blank index.html file but it should come default with modx installation. 

Google search for "index of /core/packages"

https://www.google.co.in/search?q=index+of+/core/packages/&ie=utf-8&oe=utf-8&gws_rd=cr&ei=rqyQVYCFLYXGuASylYLwCQ

A default robot.txt would be good. 

# Robots.txt
User-agent: *
Disallow: /connectors/
Disallow: /core/
Disallow: /manager/
Disallow: /setup/

**Comment by Mark-H** [comment]
I like the idea of a default robots.txt, though it is also a good way for nasty people to find out what they shouldn't be looking at. Given the number of results on that search, making them harder to find in the first place probably outweighs that though..

**Comment by lektriq** [comment]
What about renaming the ht.access file in /core to .htaccess when installing?

We already have a way to deal with this on Apache at least.

    IndexIgnore */*
    
    <Files *.php>
        Order Deny,Allow
        Deny from all
    </Files>


While the results list on a search are not great at just under 7000 results (search with quotes for exact matches), it’s not as bad as the 33,10,000 results shown using the search linked above.

**Comment by mdxthemes** [comment]
The results may be low but it displays which extra version the website is using. If any extra has any security issues, hackers might use it to gain access. Robots wont allow search engines to index but it is still visible to public. .htaccess should be placed in all such directories by default.

**Comment by OptimusCrime** [comment]
Once again, flexibility is what Modx stands for. Supply an example robots.txt. Let people use it if they will, if not it is really up to the developer/host to decide how to handle this.

**Comment by jpdevries** [comment]
Should a robots.txt also disallow assets/components? It seems like those looking to harden security would move the core folder to a non public area of there server anyways, but can't do that with assets.

**Comment by lektriq** [comment]
What about giving an option to rename /core/ht.access if it is not already at install time?

I can’t think of good reasons anyone is keeping it in a web-accessible place and not using it, unless they are configuring access in the main config, are not using Apache, or are securing it by other means.

**Comment by lektriq** [comment]
Maybe the top level .htaccess could also do with having ```Options -Indexes``` added?

**Comment by amdbuilder** [comment]
I would recommend using this for the .htaccess instead, as far as I know there isn't a reason any files should be accessible.  I'd also be inclined to rename ht.access to .htaccess, or perhaps make that a setup option (let server rename it).

```
Order Allow,Deny
Deny from all
```

**Comment by Alroniks** [comment]
What about nginx? :) I think it's problem of environment, not MODX.

**Comment by amdbuilder** [comment]
NGINX:

```
location /core {
        deny all;
        return 404;
        }
```

**Comment by Alroniks** [comment]
@amdbuilder I know how it can be applied in nginx. I mean, that .htaccess can be added to distributive, but nginx you always should configure by hand. And that this issue not for MODX, but for webserver.

**Comment by inreti** [comment]
Let me cite another popular open source project here: 

_"However, as software provider we have to take our users’ security very seriously and thus automatically help administrators as much as possible to secure their instance."_ Period. 

Conclusio: should be .htaccess in there by default. Concerning flexibility this is then an *opt-out* instead of an *opt-in* (as it is currently). If someone decides to customize that file, this guy is in charge of taking care that updates do not overwrite it... 
_robots.txt_ does not enhance any security, instead allows for fingerprinting which is one of the first steps to break security. Therefore imho even an example robots.txt should not be provided, as this may lead unexperienced users to use it without having understood the consequences.

**Comment by amdbuilder** [comment]
@Alroniks I wasn't trying to suggest you didn't know how, granted you make a valid point in that we should address other servers (nginx, lighttpd, cheroke, h2o, etc) where possible.  The simplest approach would be adding basic documentation to the rtfm site.

@inreti Great quote! 

The question now becomes what approach should MODX consider taking to improve the security?  In my mind there are a couple options:

1. Distribute MODX with a .htaccess file instead of ht.access.  Apache/Litespeed should respect the file, and all other systems should ignore it.
2. Add a "Secure Core" setting to rename the ht.access file to .htaccess.  This could provide a false sense of security should Apache/Litespeed not be in use.
3. Add server detection logic, and offer the earlier setting to Apache/Litespeed only.  Provide link to rtfm article on blocking core access for other systems.

The next question is what level of blocking should be used, blocking all access to the folder (per my earlier .htaccess example) or only blocking .php files per earlier suggestions?

My recommendation would be distributing the .htaccess file, and block access to all files.  This approach will also distribute nicely with automatic installers, increasing security across the board.

**Comment by inreti** [comment]
Definitely blocking all files. Error log, changelog etc. are not for the public. And nothing else in there too!

**Comment by OptimusCrime** [comment]
This shuold be solved in #12504.

**Comment by inreti** [comment]
Not finally solved for me. There are still open points to discuss, because #12504 does not automatically secure the install and could be to complicated for or ignored by unexperienced users (even if the rtfm would include a lot of text for that case). Let's run through some options here:

* doing nothing by default and just rely on the configcheck (my opinion: *no*)
* set up a detailed rtfm page for exactly this (should be done in any case!)
* provide a /core/.htaccess file with secure default settings for apache (and so for nginx)
* offer (default: yes) at setup (new installation) the generation of a proper .htaccess file if on apache (same for other servers) 

IMHO the last one is a good solution: every _new setup will be secured_. _Updates won't overwrite_ the files then. If we offer this #12504 is needed as well: for old installations and for cases when the file has been accidently removed or changed in an insecure way.

In general the setup should offer to secure a new installation as much at possible at this time. This could e.g. include a randomly choosed name for the manager... But that's for another issue.

**Comment by OptimusCrime** [comment]
I think you overcomplicate this a bit. I mean, this is MODX, MODX has always focused on letting the user do whatever and allowing more or less anything.

Why would the configcheck not be enough to inform the user that there is something wrong? In addition to a informative rtfm page, it should be no problem to solve this issue.

Consider this, there are thousands of Revo installations out there without the proper security stuff for the core directory. The MODX core itself is secure, highly secure. It is not like this is a huge security hole that was just discovered.

**Comment by OptimusCrime** [comment]
Most applications/CMF/CMSes like MODX does not provide any security options like this at all. This is something special that MODX just recently introduced.

Most applications just inform people to secure their webserver according the their own wishes and preferences.

**Comment by inreti** [comment]
> MODX has always focused on letting the user do whatever and allowing more or less anything.

That does not speak against choosing wise default settings and actions. Doing things right per default does not prevent you from doing whatever you want afterwards...

> The MODX core itself is secure, highly secure.

If that would be a valid argument, why do we advise to move the core outside of the webroot? 
And even if we consider that for the pure core files, as long as /cache/ and /components (!) are stored inside the core directory, MODX is in charge of securing that too, regardless of where the core is placed. It's not always about hacking something, but also about data privacy.

Besides - *nothing* should be considered *secure by default*. Don't get me wrong, that is no criticism about the core or MODX at all. It is a paradigm.

> Most applications/CMF/CMSes like MODX does not provide any security options like this at all.

Then we should not take them as an archetype... And: that is not true in general. Typo3, OwnCloud, Drupal, Piwik. They are all delivered with default .htaccess files for their core directory or - e.g. Piwik - automatically generate .htaccess files (all with contents quite the same as in PR #12503).

Why shouldn't MODX auto-provide such a file for the core directory at setup? Why would we force a user to accomplish an extra task after installation when it can be done easily (that is not complicated) during setup? It is exactly the same as deleting the setup folder after setup - it is done by default. And if not, the configcheck warns you about it...

**Comment by Mark-H** [comment]
With #12504 to warn people their core is left web accessible, and the #12503 pull will provide users a solid default they can enable by renaming a single file.

MODX targets primarily people that know what they're doing (i.e. web designers/developers and agencies) and I think giving them those two tools (a warning and an easy solution) is all that they will need.

**Comment by Mark-H** [comment]
Closing per fix in #12503
